### PR TITLE
Add allowed-plugin param to enable plugins allow-list

### PR DESCRIPTION
**Pull request by jakubm-canva**
Protection from malicious plugins - An allowed-plugins command line param accepting an allow-list of regular expression strings. e.g.: ^github.com/buildkite/.*$. If the param is used and a job arrives with BUILDKITE_PLUGINS containing a plugin that doesn't match an entry in the list then the job is rejected.

Work is a part of the https://github.com/buildkite/agent/issues/2368

I made some big git mistake in the https://github.com/buildkite/agent/pull/2394... 🤦

### CODE REVIEW

**Comment by jakubm-canva** [comment]
@moskyb @david-poirier - As per the description - I made some big damage to my old PR while "attempting to" rebase and push. PTAL here. All comments from the past are resolved and regexp is added to both allow-repositories and allow-plugins.

File: clicommand/agent_start.go [hunk 1]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
 			l.Info("Agents will disconnect after %d seconds of inactivity", agentConf.DisconnectAfterIdleTimeout)
 		}
 
+		if len(cfg.AllowedRepositories) > 0 {
+			agentConf.AllowedRepositories = []*regexp.Regexp{}
```
Comment (moskyb) [inline clicommand/agent_start.go:954]: you could probably change this for 
```suggestion
			agentConf.AllowedRepositories = make([]*regexp.Regexp, 0, len(cfg.AllowedRepositories)
```
to save a couple of slice allocations, but it's not a big deal - the length is likely to be pretty short most of the time.

File: clicommand/agent_start.go [hunk 2]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
...
+			l.Info("Repositories allow list is enabled.")
+			for _, v := range cfg.AllowedRepositories {
+				r, err := regexp.Compile(v)
+				if err != nil {
+					l.Fatal("Regex %s is allowed-repositories failed to compile: %v", v, err)
```
Comment (moskyb) [inline clicommand/agent_start.go:959]:
```suggestion
					l.Fatal("Regex %s in allowed-repositories failed to compile: %v", v, err)
```

File: clicommand/agent_start.go [hunk 3]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
...
+			}
+		}
+
+		if len(cfg.AllowedPlugins) > 0 {
+			agentConf.AllowedPlugins = []*regexp.Regexp{}
+			l.Info("Plugins allow list is enabled.")
+			for _, v := range cfg.AllowedPlugins {
+				r, err := regexp.Compile(v)
+				if err != nil {
+					l.Fatal("Regex %s is allowed-plugins failed to compile: %v", v, err)
```
Comment (moskyb) [inline clicommand/agent_start.go:971]:
```suggestion
					l.Fatal("Regex %s in allowed-plugins failed to compile: %v", v, err)
```

File: clicommand/agent_start.go [hunk 4]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
...
+				r, err := regexp.Compile(v)
+				if err != nil {
+					l.Fatal("Regex %s is allowed-repositories failed to compile: %v", v, err)
+				}
+				agentConf.AllowedRepositories = append(agentConf.AllowedRepositories, r)
+			}
+		}
+
+		if len(cfg.AllowedPlugins) > 0 {
+			agentConf.AllowedPlugins = []*regexp.Regexp{}
```
Comment (moskyb) [inline clicommand/agent_start.go:966]:
```suggestion
			agentConf.AllowedPlugins = make([]*regexp.Regexp, 0, len(cfg.AllowedPlugins)
```
see above

File: clicommand/agent_start.go [hunk 5]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
 			l.Info("Agents will disconnect after %d seconds of inactivity", agentConf.DisconnectAfterIdleTimeout)
 		}
 
+		if len(cfg.AllowedRepositories) > 0 {
+			agentConf.AllowedRepositories = []*regexp.Regexp{}
+			l.Info("Repositories allow list is enabled.")
```
Comment (moskyb) [inline clicommand/agent_start.go:955]: i think we can probably get rid of this log, it's not saying much that's useful, and we don't really have a pattern of logs when features are enabled

```suggestion
```

File: clicommand/agent_start.go [hunk 6]
```diff
@@ -939,6 +950,30 @@ var AgentStartCommand = cli.Command{
...
+				if err != nil {
+					l.Fatal("Regex %s is allowed-repositories failed to compile: %v", v, err)
+				}
+				agentConf.AllowedRepositories = append(agentConf.AllowedRepositories, r)
+			}
+		}
+
+		if len(cfg.AllowedPlugins) > 0 {
+			agentConf.AllowedPlugins = []*regexp.Regexp{}
+			l.Info("Plugins allow list is enabled.")
```
Comment (moskyb) [inline clicommand/agent_start.go:967]:
```suggestion
```

see above

**Review by moskyb — CHANGES_REQUESTED** [review]
okay, just a couple of tiny copy changes, then this bad boy is ready to go i think. thanks for sticking with it @jakubm-canva!

**Comment by jakubm-canva** [comment]
Awesome. Comments applied.

**Review by moskyb — APPROVED** [review]
awesome work 🎉

**Pull request merged**